### ISSUE TITLE
How should file content loss be recovered or how can it be avoided in the future?

### ISSUE THREAD

**Issue by MILKMILKKING** [zsviczian/obsidian-excalidraw-plugin#533]
I used Excalidraw to record the meeting minutes this morning, but when I reviewed the notes in the afternoon, I found that only the first part of the notes were saved, but not the second half.
My computer system is windows 11, Obsidian version 0.14, and I use onedrive to synchronize. The autosave and Interval for autosave options are turned on in the settings.
I tried to use Obsidian's file recovery function, but I found that the second half of the notes were not saved, so I could not recover them. I also tried onedrive's file recovery and the situation is the same.
It seems that the files are not recoverable, but I would like to try to ask if there is still a way to recover them?
I would also like to ask how should I avoid saving failure in the future?
It's a great plugin, but today's note content loss gave me a bit of a scare.

**Comment by MILKMILKKING** [comment]
I checked today's notes and remembered that I deleted the .excalidraw suffix when I created the new notes, will this action affect the saving of the notes? 
But I remember that I deleted it when I created new notes at the beginning of the day, and still saved the first half part of the notes content.

**Comment by zsviczian** [comment]
This is very concerning! ... and I don't know what to say.

I've invested lots of effort over the past weeks to address various issues regarding saving. I thought I had it properly under control.

To start with the easy one: I am sure the filename has no effect on the issue.

To make sure I understand you correctly, the problem was not that the file was compromised (i.e. it did not load due to an error), but that it simply disappeared and you could only find an earlier version in file recovery?

**Comment by zsviczian** [comment]
Here's an idea to try to find an earlier version of the file. OneDrive sync maintains a history of your files.  Check to see if it has saved an earlier version.

![image](https://user-images.githubusercontent.com/14358394/158865622-091f4e02-40f3-42d8-bfdc-8cb7c5c00f05.png)


To be on the safe side, you might consider these two things:
1) Manually press the save button before closing Excalidraw.  
- I've implemented some logic to limit unnecessary saving of the file. This is to ensure that Excalidraw works well with sync and to avoid your work being overwritten by an older version from a different device when you switch devices during your work. (e.g. yesterday I took my meeting notes in Excalidraw on my iPad, then I processed those notes at my desktop, and finally, at a later meeting I made some further edits on my iPad). I could imagine that for some reason autosave did not detect the drawing has changed... 
  - question: as you were drawing did you see the disk Icon turning red and back? If the disk Icon is not turning red, then it is not recognizing that you've made changes to the file. If it is not turning white again, it stopped saving.
- I will implement an enhancement for Excalidraw to display a short notice when you hit autosave, to confirm that saving was successful.

2) avoid terminating Obsidian with an Excalidraw drawing open. When Obsidian is terminated, Obisidian does not guarantee that it will wait for the plugins to save data. This was one of the key sources of corrupted Excalidraw files in the past. To avoid corrupted files, when Obsidian is terminated, I don't attempt to save, your unsaved data (since the last autosave) is lost. But even if you manually save your data right before terminating Obsidian, the next time you open Obsidian it could result in a version conflict (as explained in point 1). So my practice is closing Excalidraw before closing Obsidian.

**Comment by MILKMILKKING** [comment]
First of all I would like to say that the plugin you designed is great! Thank you for putting a lot of time and effort into optimizing her!
The image below shows the file history of onedrive, I restored each version and still only have the notes the first few minutes recorded.
![image](https://user-images.githubusercontent.com/69726213/158919992-e4331faf-b4f1-4810-99d6-7c461d96d9e9.png)
The image below is my notes after restoration, the file history of onedrive basically stays at these contents, but in reality I recorded much more than that. I was under the impression that these contents were recorded a few minutes after the new notes were created.
![image](https://user-images.githubusercontent.com/69726213/158920647-f4823cc4-73b8-4e90-a173-22b2c3ffc6d7.png)
The new notes were created yesterday at 9:30AM BST and lasted for an hour; I guess excalidraw only saved them in the first few minutes. Since I made quite a few changes in other OBSIDIAN notes during this hour, the changes in these other notes were saved. Maybe we can rule out the problem of obsidian software saving.

In order to avoid the problem of content loss in the future, could you add a new note save time in the bottom left corner of excalidraw? Something like this.
![image](https://user-images.githubusercontent.com/69726213/158929094-078db325-6d9f-4db2-af36-ee85dd034abb.png)

**Comment by MILKMILKKING** [comment]
Or can it be placed at the bottom of the obsidian software there? (to avoid too much obscuring of excalidraw content)
Mainly, I want to have a record of the time saved to confirm whether the notes are saved successfully or not for some particularly important notes.
Thanks again for your great work on excalidraw!
![image](https://user-images.githubusercontent.com/69726213/158929921-3d1b0d09-8af1-4d63-9381-57788c654e89.png)

**Comment by zsviczian** [comment]
I will think about this. Displaying the last save time in the status bar is a possible solution. Currently the save button is red when there are unsaved changes.

**Comment by MILKMILKKING** [comment]
I just witnessed the sudden disappearance of the content, which I guess was related to a plugin I was using.
I recently downloaded a plugin for hovering preview boxes called Hover Editor, and I just created a new excalidraw note and locked the page; after about a while, I hovered the note with Hover Editor and then continued to record the content. At this point I noticed that the save button was always red, and would not turn white no matter how I clicked on it. I looked at the onedrive save log, and the onedrive update log shows that this excalidraw note was last saved 5 minutes ago, which I assume is the time I hovered the window.
At this point I close this note and the content of this 5 minute record disappears.
So I presume it's a plugin conflict problem.

**Comment by MILKMILKKING** [comment]
> I will think about this. Displaying the last save time in the status bar is a possible solution. Currently the save button is red when there are unsaved changes.

Please ignore my request to display the save time. I think the color change of the save button is enough to indicate whether the save was successful or not. I found today, after repeated tests, that as long as the save button turns white, the save is successful.
So I don't think I need to spend time designing the request for the save time.
Thank you again for designing the excalidraw piece of the plugin, it's fantastic!
Also, I was unable to get a coffee for you and the following prompt appeared when I clicked the button.
![image](https://user-images.githubusercontent.com/69726213/158955510-22aa92c8-bcf5-414b-ab08-67fbe45036b9.png)

**Comment by zsviczian** [comment]
Thanks. Hopefully this is just a temporary thing with ko-fi. I just tried, and it worked for me... but there could be a location specific issue as well.

As for the issue with saving, I really hope this will not happen again. If it does, we will need to investigate if that has anything to do with onedrive. I used onedrive with Obsidian for a few month, but then I purchased Obsidian sync, especially because I use Obsidian on an iPad, an Android phone, and two different windows PCs, and OneDrive was causing me problems with versions. Since moving to sync, I've had a very smooth experience.
